### ISSUE TITLE
aligned_alloc needs -mmacosx-version-min=10.15 to work

### ISSUE THREAD

**Issue by cjappl** [realtime-sanitizer/radsan#3]
Transferring this issue here from the discord so we can talk about it and save things for posterity. The original post was:

> One issue I discovered was that I couldn't just build it for my version of macos (Monterey) out of the box - I had to turn off IOS support using -DCOMPILER_RT_ENABLE_IOS=OFF, otherwise I got this error while building clang/llvm: 

```
clang: error: invalid argument '-mmacosx-version-min=10.15' not allowed with '-mios-simulator-version-min=9.0'
clang: warning: using sysroot for 'iPhoneSimulator' but targeting 'MacOSX' [-Wincompatible-sysroot]
```

The (truncated by me) response from @davidtrevelyan was:

> We originally hacked that -mmacosx-version-min=10.15 flag into the compile flags for the RADSan part of compiler-rt, because we wanted to intercept functions (like aligned_alloc) that were only introduced in macOS 10.15, and the rest of compiler-rt is hard-limited to macOS 10.10. Without this hack, we can't intercept everything we want to. It worked kind-of OK for macOS, but we never actually tried building RADSan for iOS or iPhone Simulator.

> If we remove this hack and stick with the deployment target 10.10 of compiler-rt and clang, we can't think of a way to intercept these newer system calls, which we think would put a little bit of a hole in the safety net we're trying to build. Here's what we think the options are:
> 1. Don't build RADSan for iOS or iPhoneSimulator at all (i.e. what you do with COMPILER_RT_ENABLE_IOS)
> 2. Build RADSan for all Apple platforms, but remove all interceptors for functions that had been introduced after macOS 10.10
> 3. Try to tease RADSan away from clang and compiler-rt, and build it as a separate library to remove this deployment target limitation altogether.

**Comment by cjappl** [comment]
One thing I discovered while rooting around is there seems to be cut outs for specific calls like this. Specifically, `aligned_alloc` seems to be `#ifdef`'d in a number of other sanitizers. [See asan on linux](https://github.com/llvm/llvm-project/blob/ff485a0e77a55847cb50768b01c04fe45a6879ea/compiler-rt/lib/asan/asan_malloc_linux.cpp#L111-L116):

```
#if SANITIZER_INTERCEPT_ALIGNED_ALLOC
INTERCEPTOR(void*, aligned_alloc, uptr boundary, uptr size) {
  GET_STACK_TRACE_MALLOC;
  return asan_aligned_alloc(boundary, size, &stack);
}
#endif // SANITIZER_INTERCEPT_ALIGNED_ALLOC
```

The mac sanitizer "malloc" [include file](https://github.com/llvm/llvm-project/blob/main/compiler-rt/lib/sanitizer_common/sanitizer_malloc_mac.inc) does not mention aligned_alloc at all.

And the `SANITIZER_INTERCEPT_ALIGNED_ALLOC` define is explicitly "if this is not a mac":

```
compiler-rt/lib/sanitizer_common/sanitizer_platform_interceptors.h
495:#define SANITIZER_INTERCEPT_ALIGNED_ALLOC (!SI_MAC)
```

It seems like other sanitizers also have this problem, so we could possibly open a bug with clang and see if they have any bright solutions here. In the meantime, I think there could be a case made to just omit aligned_alloc from the set right now.

This is not 100% perfect, as described in the discord it is "cutting holes in the safety net", but I think there is an opportunity to not let the perfect and good be enemies here, especially if there is the precedent of other sanitizers ignoring it. There is also the possibility to add this in the future for this sanitizer and ALL sanitizers if the fix is done properly

**Comment by cjappl** [comment]
I would vouch for putting the `aligned_alloc` interceptor behind an `#ifdef SANITIZER_INTERCEPT_ALIGNED_ALLOC` for now, which would allow it to continue to exist on Linux, and not fail the build on Darwin. Eventually when they move the minimum version up from 10.10 to 10.15 for the sanitizers, we would get `aligned_alloc` for free.

I have done this in a PR here for demonstration purposes:
https://github.com/realtime-sanitizer/llvm-project/pull/1


It seems like the change to the llvm project ([here](https://github.com/llvm/llvm-project/commit/259ad9cf08c97f9ea977d07c241e6c97bcb3f189)) was done in 2017, before aligned_malloc was released in osx 10.15 in 2019. I think this is a completely valid change to make to llvm, we could attempt to upstream it.

The initial bump to 10.10 as the minimum version was done [here](https://github.com/llvm/llvm-project/commit/41c52889b90ded563e1aa56cfa7764e083323b00), and [here](https://github.com/llvm/llvm-project/commit/b87fc09dceeff9cb838884961d817c0ab0257176). They bumped the version when the OS version was 5 years old, which 10.15 is 4 years old currently. 

Through the course of this, I rediscovered what @davidtrevelyan said in the discord:

> Regarding the SDK version and #ifdef guards: so, you're right - you can indeed do this sort of check. The issue is that the check must be done on the deployment target (i.e. the minimum supported version of macOS). The deployment target is, of course, usually older than the SDK you're using to compile. LLVM, unsurprisingly, is quite restricted in terms of the deployment target, because they need to support macOS all the way back to 10.10. In fact, as a safety mechanism, they actually override the CMake configuration variable that is normally used, and they print out a warning to explain why if you try to set it to something too new. That's unfortunately where the issue is - the deployment target is fixed - and it's why our "solution" is a total hack job. 

Seems like you all already thought of this somewhat 😅

**Comment by davidtrevelyan** [comment]
Great investigations @cjappl! Thanks for diving into the other sanitisers to see what they do - that's super helpful.

I agree with your conclusion. I think the best compromise for now is to put the `aligned_alloc` interceptor behind a preprocessor guard that checks for 10.15. Funnily enough this has kind of come full circle - a couple of weeks ago I actually drafted this change as well - I've thrown up a quick draft PR [here](https://github.com/realtime-sanitizer/llvm-project/pull/3) for you to check it out. It pretty much does the same things that you did, but mine is less elegant than yours. I prefer your way of organising the preprocessor definitions. 

Here are the key differences we need to be aware of: 

1. In the tests, I also guarded use of `std::shared_mutex` with a check for macOS 10.12,
2. My check for 10.15 was based on `__MAC_OS_X_VERSION_MIN_REQUIRED >= __MAC_10_15`, which is slightly different to your check, which is based on `__ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__ < 101500` - we will just need to double check whether this matters or not,
3. I forgot to guard `INTERCEPT_FUNCTION(aligned_alloc);`, and
4. I didn't change the `DEFAULT_SANITIZER_MIN_OSX_VERSION` to 10.15. 

I reckon the best approach here would be to make our decisions and then create a new PR with the elements we like best from both existing drafts? Is that something you'd be interested in doing? No problem if not, I'm happy to do it as well.

Here are my initial thoughts on the decisions we need to make (based on the differences above):

1. I think we should keep the `std::shared_mutex` test if possible, and I think we should treat it in the same way as we treat the versioning for `aligned_alloc`. This probably means keeping the 10.12 guard?
2. Choosing the right check should be easy here, just needs a bit of reading
3. I like your `RADSAN_MAYBE_INTERCEPT_ALIGNED_ALLOC` - let's use that, or perhaps even (if it's not too much effort) we could add a more reusable `#define RADSAN_MAYBE_INTERCEPT(interceptee_function, preprocessor_predicate)`? 
4. My vote would be for not changing the default deployment target here. I think having an inconsistency with the rest of the compiler components is asking for trouble. I also wonder; if the deployment target of the compiler is so low, are the C++ runtime libraries on user systems likely to be using newer functions like `aligned_alloc` anyway? And, as you said above, if we add our preprocessor guards in a way that is independent of the compiler-rt sanitizer configuration, then we will just activate them "for free" when LLVM bump their deployment target.

So the problem with my thinking for 4. is that we then can't run tests for `shared_mutex` and `aligned_alloc`, etc. Maybe we could get around this by making an override for `DEFAULT_SANITIZER_MIN_OSX_VERSION` configurable at CMake time? This might be playing with fire a bit, but it would allow us to test this stuff on our dev machines without too much effort.  

What do you think, @cjappl?

**Comment by davidtrevelyan** [comment]
Sounds great, @cjappl ! I agree on your points - thanks for flagging the existing styles/prior art in other sanitizers.

Here's where I think we are with the decision making:

1. We'll protect the `shared_mutex` test with a guard for 10.12
2. We'll use the preprocessor def `__ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__` to check for the deployment target
3. We'll keep the maybe-intercept preprocessor definitions simple and consistent with the rest of the sanitizers, and go with `RADSAN_MAYBE_INTERCEPT_ALIGNED_ALLOC` style.
4. We won't change `DEFAULT_SANITIZER_MIN_OSX_VERSION`, unless the user has explicitly passed in a RADSan-specific CMake option to override it.

Seems sensible to me! Thanks for taking this on - appreciate your support :)

**Comment by cjappl** [comment]
I'll get it done, thanks for the trust

**Comment by cjappl** [comment]
First attempt here:

https://github.com/realtime-sanitizer/llvm-project/pull/4

Looking forward to continuing the discussion over there :)

**Comment by davidtrevelyan** [comment]
Thanks to @cjappl, this should now be fixed after updating the LLVM fork to [beb2b62](https://github.com/realtime-sanitizer/radsan/commit/beb2b62d2f1d6783713175bf0ee6db33fe763380). Many thanks! Closing this now.
